include/LightGBM/c_api.h

```diff
@@ -214,6 +214,17 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetCreateFromCSC(const void* col_ptr,
                                                 const DatasetHandle reference,
                                                 DatasetHandle* out);
 
+/*!
+ * \brief Create sample indices for total nrow.
+ * \param total_nrow Number of all data rows
+ * \param parameters Additional parameters, specify sample count and random seed in parameter
+ * \param[out] out Created indices, type is int32_t, caller should insure out contains enough space to hold indices
+ * \return 0 when succeed, -1 when failure happens
+ */
+LIGHTGBM_C_EXPORT int LGBM_SampleIndices(int32_t total_nrow,
```

Comment:
I think it will be useful to return actual number of sampled rows (`int32_t* out_len`) as well as indices themselves. Caller can call this function first time to get number of elements, allocate memory, and call this function second time to get indices. Something similar to allocating string buffers.
https://github.com/microsoft/LightGBM/blob/c738c83bbddc80808b47d0603e182e9595b7b2f7/python-package/lightgbm/basic.py#L1878-L1900

Or we can define a separate helper function for calculating number of indices given particular config params.

I'm suggesting this because I don't like that we have multiple places of the following duplicated code:
```
int sample_cnt = static_cast<int>(total_nrow < config.bin_construct_sample_cnt ? total_nrow : config.bin_construct_sample_cnt);
```
at both Python and cpp sides.

Reply:
The key problem here is that we have to implement some logic in Python and make the behavior the same with cpp side.

Feature request #4321 can help sync default values in Python and cpp side. But it's not feasible to completely remove logic duplication in both side. In this PR, the sampling data from multiple file logic can't be easily implemented in cpp side.

I'm going to create a function to get sample count in Python side to remove duplication. For cpp side, I separate PR is more appropriate.

Reply:
Commit 96ebbfe7 adds a separate function `_get_sample_count` in `basic.py` to avoid code duplicate in Python side.

Reply:
Thanks! But why we can't add a separate function **at cpp side** and call it from Python in `_get_sample_count()` function? So that computation of `sample_cnt` will be performed only in one place - at cpp side.

Reply:
Then we'll be able to do something similar to
https://github.com/microsoft/LightGBM/blob/08d1ce4bbf9391f5cc945f5f015840f4a57774a6/include/LightGBM/c_api.h#L637

Reply:
Just looked at `c_api.cpp` and I've got a few questions:

1. Is it find to use one API (e.g. `LGBM_SampleCount`) inside other APIs? This is to remove code duplicate in cpp side.
2. If it's OK to the 1st question, let's add API `LGBM_SampleCount` first, then send a separate PR to remove code duplicate in cpp side?

This removes the need of syncing default sample count, but still leaves `ZERO_THRESHOLD`.

Reply:
1. TBH, I don't know. But we can create internal function and call it from both (`LGBM_SampleCount` and `LGBM_SampleIndices`) C API functions if it will be a problem.
2. OK, agree with you about the follow-up PR to not overcomplicate this PR and its' discussion.

> This removes the need of syncing default sample count, but still leaves `ZERO_THRESHOLD`.

OK, no problem! My suggestion was to remove duplicated **code**. For duplicated definitions of constants you've alredy created a separate issue - thanks a lot.

Reply:
C API `LGBM_SampleCount` added.